File: problems/problems_3106/Solution.java
package problems.problems_3106;

import com.alibaba.fastjson.JSON;
import java.util.*;
import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    private int distance(char a, char b) {
        return Math.min((a - b + 26) % 26, (b - a + 26) % 26);
    }
    public String getSmallestString(String s, int k) {
        StringBuilder sb = new StringBuilder();
        int idx = 0;
        while (idx < s.length() && k > 0) {
            if (s.charAt(idx) == 'a') {
                sb.append('a');
            } else {
                int d = distance('a', s.charAt(idx));
                if (d <= k) {
                    sb.append('a');
                    k -= d;
                } else {
                    sb.append((char) (s.charAt(idx) - k));
                    k = 0;
                }
            }
            idx++;
        }
        sb.append(s.substring(idx));
        return sb.toString();
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        String s = jsonStringToString(inputJsonValues[0]);
		int k = Integer.parseInt(inputJsonValues[1]);
        return JSON.toJSON(getSmallestString(s, k));
    }
}

File: problems/problems_3106/solution.ts
function getSmallestString(s: string, k: number): string {
    const distance = (a: string, b: string) => {
        return Math.min((b.charCodeAt(0) - a.charCodeAt(0) + 26) % 26, (a.charCodeAt(0) - b.charCodeAt(0) + 26) % 26);
    }
	let idx: number = 0;
	const result: string[] = [];
	while (idx < s.length && k > 0) {
		if (s.charAt(idx) === 'a') {
			result.push('a');
		} else {
			const d: number = distance('a', s.charAt(idx));
			if (d <= k) {
				result.push('a');
				k -= d;
			} else {
				result.push(String.fromCharCode(s.charCodeAt(idx) - k));
				k = 0;
			}
		}
		idx++;
	}
	result.push(s.substring(idx));
	return result.join('');
};

export function Solve(inputJsonElement: string): any {
    const inputValues: string[] = inputJsonElement.split("\n");
    const s: string = JSON.parse(inputValues[0]);
    const k: number = JSON.parse(inputValues[1]);
    return getSmallestString(s, k);
}
